### PR TITLE
Fix `monomial_coefficients` for submodules of free modules from sage.modules

### LINKED ISSUE #34455: Fix monomial_coefficients for submodules of free modules from sage.modules
<div id="comment:0"></div>

Bug: 

```
sage: V = ZZ^3
sage: U = V.submodule([[1, 2, 3], [1, 1, 1]])
sage: U
sage: u = U([2, 3, 4])
sage: u.monomial_coefficients()
{0: 2, 1: 3, 2: 4}
```
This is incorrect - by definition of `monomial_coefficients`, basis indices are mapped to coefficients; but

```
sage: u.parent().basis()
[
(1, 0, -1),
(0, 1, 2)
]
```

We also add a method `_test_monomial_coefficients` to run some tests on this and related methods of `ModulesWithBasis`.

Part of #30309

CC:  @tscrim @mmasdeu

Component: **linear algebra**

Author: **Matthias Koeppe, ...**

_Issue created by migration from https://trac.sagemath.org/ticket/34455_

### PR DESCRIPTION
**Pull request by mkoeppe**
- Rebased version of #34455
- Part of https://github.com/sagemath/sage/issues/30309

Fixes #34455

### :memo: Checklist

- [x] The title is concise and informative.
- [ ] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/categories/modules_with_basis.py [hunk 1]
```diff
@@ -1467,6 +1467,39 @@ def monomial_coefficients(self, copy=True):
...
+            See the documentation for :class:`TestSuite` for more information.
+            """
+            tester = self._tester(**options)
+            base_ring = self.parent().base_ring()
+            basis = self.parent().basis()
+            try:
+                d = self.monomial_coefficients()
+            except NotImplementedError:
+                return
+            tester.assertTrue(all(value.parent() is base_ring
```
Comment (tscrim) [inline src/sage/categories/modules_with_basis.py]:
```suggestion
            tester.assertTrue(all(value.parent() == base_ring
```
The base ring might not be a `UniqueRepresentation`.

File: src/sage/modules/free_module_element.pyx [hunk 2]
```diff
@@ -949,6 +949,51 @@ cdef class FreeModuleElement(Vector):   # abstract base class
...
+            [ 1  0 -1]
+            [ 0  1  2]
+            sage: u = U([2, 3, 4])
+            sage: u.monomial_coefficients()
+            {0: 2, 1: 3}
+        """
+        base_ring = self.parent().base_ring()
+        if self.parent().is_ambient() and base_ring == self.parent().coordinate_ring():
+            return self.dict(copy=copy)
+        else:
```
Comment (tscrim) [inline src/sage/modules/free_module_element.pyx]: I find `else:` statements like this unnecessary and just adding extra (unnecessary) indentation.

**Review by tscrim — CHANGES_REQUESTED** [review]

File: src/sage/modules/free_module_element.pyx (hunk 2 above)
Comment (mkoeppe) [reply src/sage/modules/free_module_element.pyx]: I agree, done in 2eb5de8fd64

**Review by tscrim — APPROVED** [review]
Okay, LGTM.

**Comment by mkoeppe** [comment]
Thank you!

**Pull request merged**